### PR TITLE
BlockId removal: refactor: StorageProvider

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of `Backend::StorageProvider` trait from:
block: `BlockId<Block>` to: hash: `&Block::Hash`

Companion for: paritytech/substrate#12510
cumulus companion: paritytech/cumulus#1770

### CODE REVIEW

File: node/client/src/lib.rs [hunk 1]
```diff
@@ -418,66 +418,69 @@ impl sc_client_api::BlockBackend<Block> for Client {
...
 
-impl sc_client_api::StorageProvider<Block, crate::FullBackend> for Client {
+impl sc_client_api::StorageProvider<Block, crate::FullBackend> for Client
+where
+	Block: BlockT,
+{
 	fn storage(
 		&self,
-		id: &BlockId<Block>,
+		hash: &<Block as BlockT>::Hash,
```
Comment (bkchr) [inline node/client/src/lib.rs:427]:
```suggestion
		hash: &Block::Hash,
```

You don't need to fully qualify the trait.

**Review by bkchr — APPROVED** [review]

**Review by altonen — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]

**Review by skunert — APPROVED** [review]

File: node/client/src/lib.rs (hunk 1 above)
Comment (michalkucharczyk) [reply node/client/src/lib.rs:427]: This won't compile:
```error[E0223]: ambiguous associated type
   --> node/client/src/lib.rs:427:10
    |
427 |         hash: &Block::Hash,
    |                ^^^^^^^^^^^ help: use fully-qualified syntax: `<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic> as Trait>::Hash`
```

File: node/client/src/lib.rs [hunk 2]
```diff
@@ -418,66 +418,69 @@ impl sc_client_api::BlockBackend<Block> for Client {
 	}
 }
 
-impl sc_client_api::StorageProvider<Block, crate::FullBackend> for Client {
+impl sc_client_api::StorageProvider<Block, crate::FullBackend> for Client
+where
+	Block: BlockT,
+{
```
Comment (bkchr) [inline node/client/src/lib.rs:424]:
```suggestion
impl sc_client_api::StorageProvider<Block, crate::FullBackend> for Client {
```

File: node/client/src/lib.rs (hunk 1 above)
Comment (bkchr) [reply node/client/src/lib.rs:427]: Ahh yeah, I thought `Block` is a generic type here. However, it is the actual type. So, this doesn't work. Sorry!

**Pull request merged**